### ISSUE TITLE
Better/safer access of missing values in Painless

### ISSUE THREAD

**Issue by costin** [elastic/elasticsearch#33816]
**Describe the feature**:

Currently in master, the correct way to access a doc value is to do a check on size ( see #32207) to avoid an exception when dealing with missing values: `doc['field'].size() > 0`

However our examples do not use this form (though the docs mentions that in a paragraph).
Without this check, a document with missing fields can cause a working Painless script to throw an exception which is unexpected.
As discussed in the search/aggs meeting it would be desirable to have an intuitive, short and safe access to the doc value regardless of the existence of the value.

Maybe related to #25913 /cc @nik9000

**Comment by rjernst** [comment]
There are 2 types of existence checks that need to be supported. First is whether a mapping exists. This is what checking for a field in `doc` does today. Second is whether the current document has any documents. This is what the size() check does. What changed in #32207 is what happens when asking for `.value` when no values exist.

> it would be desirable to have an intuitive, short and safe access to the doc value regardless of the existence of the value.

What is unintuitive about checking size? It's like doing `hasNext()` on an iterator before you can call `next()`.

Also note the choice of adopting size() checks as the preferred way to check for missing values is indeed related to #25913. If we change the return value per document (eg return null for some documents), we can't lazily cache the return value, which is the plan to do only one hash lookup regardless of the number of documents the script is run on. The returned object from `doc['field']` is already advanced externally as the documents are evaluated by scripts.

**Comment by GlenRSmith** [comment]
TL;DR:
1. Removing `.value` (if I understand what that means) will break BC.
1. I'd like to work to improve the docs for scripts metrics aggs, please advise.

Prompted by a [Discuss issue](https://discuss.elastic.co/t/official-document-question/207033), I'd like to improve the public documentation around this issue, specifically in the `Script` sections of various individual Metrics Agg pages. The accessors there are typically shown "naively" (neglecting the behavior where a doc is missing the field, or when a doc has multiple values for the field), e.g.
```
"script" :   {"source" : "doc.grade.value" }
```

My initial intention was to propose an additional section below each of these that mentions the error returned by that query when a doc is missing the value, and show a guarded form that matches the error, e.g.

```
"script" : {
  "source" : "if (doc['grade'].size()==0) { return null; } return doc.age.value;"
}
```

The Discuss issue mentioned that omitting `.value` remediated the problem, and I was able to confirm this - except that it has the consequence of causing Metrics aggs to behave differently for documents with multiple values on the field.
In 7.4, using `.value` causes only the first value of the field in each doc to be included in the calculation of the metric, where omitting `.value`, e.g. `doc.grade` includes all values from each doc in the metric calculation.
**If the implementation is changed so that `doc.<field>` now only returns the first value, extant aggregation scripts may be disrupted.**
To be honest, though, I wonder how many people use `doc.<field>` _intentionally_ to get all values versus accidentally, .e.g. like the above user they discovered dropping `.value` made the missing value error go away.

Regarding possible changes, IMO:
- both of these scenarios (value-per-doc, all values from doc) *may* be intentional on the users' part, and should be available in future
- the choice is more obscure than is ideal; the choice of "all values from doc" might be surfaced in some explicit way.

Regarding possible current doc improvements, please advise broad strokes what would be acceptable. I want to say links to documentation for accessing fields in scripts would be helpful, but then I see _that_ doc doesn't have any references to dot operator (`doc.field`) access, nor does it distinguish between the behaviors of `doc.field` and `doc.field.value` where there is an array of values.

**Comment by jdconrad** [comment]
@GlenRSmith I completely agree this is a gap in the documentation. Common context parameters such as doc and source should at the very least have entire sections with simple examples dedicated to each. I was considering adding these sections to the Painless guide. Behavioral changes are obviously more difficult, but if it's the right thing to do we should come up with a plan for appropriate deprecations. I think the problem with obscurity is largely that users may not be aware that `doc['field']` (or `doc.field`) is actually returning what amounts to a List. `.value` was likely added a shortcut for users since most fields end up being single-valued so it wasn't as strange to constantly have to do `doc['field'][0]`.

**Comment by jdconrad** [comment]
This is covered by the planned fields api for Painless.
